plugins/ingest/src/main/java/org/elasticsearch/ingest/processor/CompoundProcessor.java

```diff
@@ -59,23 +62,32 @@ public String getType() {
 
     @Override
     public void execute(IngestDocument ingestDocument) throws Exception {
-        try {
-            for (Processor processor : processors) {
+        for (Processor processor : processors) {
+            try {
                 processor.execute(ingestDocument);
-            }
-        } catch (Exception e) {
-            if (onFailureProcessors.isEmpty()) {
-                throw e;
-            } else {
-                executeOnFailure(ingestDocument);
+            } catch (Exception e) {
+                if (onFailureProcessors.isEmpty()) {
+                    throw e;
+                } else {
+                    executeOnFailure(ingestDocument, e, processor.getType());
+                }
+                return;
             }
         }
+    }
+
+    private void resetErrorContext(Map<String, String> ingestMetadata) {
 
     }
 
-    void executeOnFailure(IngestDocument ingestDocument) throws Exception {
+    void executeOnFailure(IngestDocument ingestDocument, Exception cause, String failedProcessorType) throws Exception {
+        Map<String, String> ingestMetadata = ingestDocument.getIngestMetadata();
+        ingestMetadata.put(ON_FAILURE_MESSAGE_FIELD, cause.getMessage());
+        ingestMetadata.put(ON_FAILURE_PROCESSOR_FIELD, failedProcessorType);
```

Comment:
> Is this enough info from the error? I was expecting something more detailed like we do in ElasticsearchException#toXContent. Is that not needed?

That makes sense to do, right now we may lose a lot of details.

> One more thing, can it happen that we have multiple errors but we keep track of only one of them due to key collisions in the map?

The exception is only available in the context of the current on failure processors. They need to act upon it.